server.js

```diff
@@ -3905,6 +3906,268 @@ cache(function(data, match, sendBadge, request) {
+    } else if (!json.access_token) {
+      return end('The GitLab OAuth process did not return an access token.');
+    }
+
+    gitlabHelpers.addToken(json.access_token);
+  });
+});
+
+// GitLab stars & forks integration
+camp.route(/^\/gitlab\/((https?)\/([^/]+)\/)?(star|fork)s\/([^/]+)\/(.+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  const isStar = (match[4] === 'star');
+  const user = match[5];
+  const repo = match[6];
+  const format = match[7];
+  const apiURL = gitlabHelpers.baseURL(match) + '/api/v4/projects/' + user + '%2F' + repo;
+  const badgeData = getBadgeData((isStar ? 'stars' : 'forks'), data);
+
+  gitlabHelpers.request(request, apiURL, {}, true, function(err, res, buf) {
+    if ((err != null) || (res.statusCode !== 200)) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+    } else {
+      try {
+        const data = JSON.parse(buf);
+        badgeData.text[1] = metric(parseInt((isStar ? data.star_count : data.forks_count)));
+        badgeData.colorscheme = 'blue';
+        sendBadge(format, badgeData);
+      } catch(e) {
+        badgeData.text[1] = 'invalid';
+        sendBadge(format, badgeData);
+      }
+    }
+  });
+}));
+
+// GitLab contibutors integration
+camp.route(/^\/gitlab\/((https?)\/([^/]+)\/)?contributors\/([^/]+)\/(.+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  const user = match[4];
+  const repo = match[5];
+  const format = match[6];
+  const apiURL = gitlabHelpers.baseURL(match) + '/api/v4/projects/' + user + '%2F' + repo + '/repository/contributors';
+  const badgeData = getBadgeData('contributors', data);
+
+  gitlabHelpers.request(request, apiURL, {}, true, function(err, res, buf) {
+    if ((err != null) || (res.statusCode !== 200)) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+    } else {
+      try {
+        badgeData.text[1] = metric(parseInt(res.headers['x-total']));
+        badgeData.colorscheme = 'blue';
+        sendBadge(format, badgeData);
+      } catch(e) {
+        badgeData.text[1] = 'invalid';
+        sendBadge(format, badgeData);
+      }
+    }
+  });
+}));
+
+// GitLab build status integration
+camp.route(/^\/gitlab\/((https?)\/([^/]+)\/)?build\/([^/]+)\/([^/]+)(?:\/([^/]+))?\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  const user = match[4];
+  const repo = match[5];
+  const branch = match[6] || 'master';
+  const format = match[7];
+  const apiURL = gitlabHelpers.baseURL(match) + '/' + user + '/' + repo + '/badges/' + branch + '/pipeline.svg';
+  const badgeData = getBadgeData('build', data);
+
+  request(apiURL, function(err, res, buffer) {
+    if ((err != null) || (res.statusCode !== 200)) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+    } else {
+      xml2js.parseString(buffer.toString(), function (err, data) {
```

Comment:
Is there no way to get the response as JSON here? If not, it's not amajor issue, XML is fine as well.

Reply:
I can't remember if the original API endpoint required authorization too but either way, piggy-backing a non-API resource allows us to avoid rate limiting :)